hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/cancelairdrops/TokenCancelAirdropDecoderTest.java

```diff
@@ -0,0 +1,327 @@
+                    OWNER_ACCOUNT_AS_ADDRESS,
+                    FUNGIBLE_TOKEN_HEADLONG_ADDRESS,
+                    0L),
+            Tuple.of(
+                    asHeadlongAddress(SENDER_ID.accountNum()),
+                    OWNER_ACCOUNT_AS_ADDRESS,
+                    FUNGIBLE_TOKEN_HEADLONG_ADDRESS,
+                    0L),
+            Tuple.of(
+                    asHeadlongAddress(SENDER_ID.accountNum()),
+                    OWNER_ACCOUNT_AS_ADDRESS,
+                    FUNGIBLE_TOKEN_HEADLONG_ADDRESS,
+                    0L),
+            Tuple.of(
+                    asHeadlongAddress(SENDER_ID.accountNum()),
+                    OWNER_ACCOUNT_AS_ADDRESS,
+                    FUNGIBLE_TOKEN_HEADLONG_ADDRESS,
+                    0L),
+            Tuple.of(
+                    asHeadlongAddress(SENDER_ID.accountNum()),
+                    OWNER_ACCOUNT_AS_ADDRESS,
+                    FUNGIBLE_TOKEN_HEADLONG_ADDRESS,
+                    0L),
+            Tuple.of(
+                    asHeadlongAddress(SENDER_ID.accountNum()),
+                    OWNER_ACCOUNT_AS_ADDRESS,
+                    FUNGIBLE_TOKEN_HEADLONG_ADDRESS,
+                    0L)
+        })));
+        given(attempt.inputBytes()).willReturn(encoded.toArrayUnsafe());
+
+        assertThrows(HandleException.class, () -> subject.decodeCancelAirdrop(attempt));
+    }
+
+    @Test
+    void failsIfTokenIsNull() {
+        // given:
+        given(attempt.configuration()).willReturn(configuration);
+        given(attempt.enhancement()).willReturn(enhancement);
+        given(enhancement.nativeOperations()).willReturn(nativeOperations);
+        given(configuration.getConfigData(TokensConfig.class)).willReturn(tokensConfig);
+        given(tokensConfig.maxAllowedPendingAirdropsToCancel()).willReturn(10);
+        given(nativeOperations.getToken(FUNGIBLE_TOKEN_ID.tokenNum())).willReturn(null);
+        given(addressIdConverter.convert(asHeadlongAddress(SENDER_ID.accountNum())))
+                .willReturn(SENDER_ID);
+        given(addressIdConverter.convert(OWNER_ACCOUNT_AS_ADDRESS)).willReturn(OWNER_ID);
+
+        final var encoded = Bytes.wrapByteBuffer(CANCEL_AIRDROP.encodeCall(Tuple.singleton(new Tuple[] {
+            Tuple.of(
+                    asHeadlongAddress(SENDER_ID.accountNum()),
+                    OWNER_ACCOUNT_AS_ADDRESS,
+                    FUNGIBLE_TOKEN_HEADLONG_ADDRESS,
+                    0L)
+        })));
+        given(attempt.inputBytes()).willReturn(encoded.toArrayUnsafe());
+
+        assertThrows(
+                HandleException.class,
+                () -> subject.decodeCancelAirdrop(attempt),
+                PENDING_AIRDROP_ID_LIST_TOO_LONG.protoName());
+    }
+
+    @Test
+    void failsIfTokenIsNullHRCFT() {
+        final var encoded = Bytes.wrapByteBuffer(HRC_CANCEL_AIRDROP_FT.encodeCallWithArgs(OWNER_ACCOUNT_AS_ADDRESS));
+        given(attempt.inputBytes()).willReturn(encoded.toArrayUnsafe());
```

Comment:
I think this test isn't testing what you think it is testing, IMO.  Looking at `HtsCallAttempt.redirectTokenId` we see that it is supposed to fail (throw exception) if the call attempt is _not_ a redirect.  It is _not_ a redirect if the redirect address is null.  Our `attempt` here is a mock, there are no givens for `isRedirect` or `redirectAddress`, so you'd think it would be null, thus throw an IllegalStateException, but it doesn't.  It continues (and eventually throws a `HandleException(INVALID_TOKEN_ID)`.  But why did it continue, i.e., why did it think it _was_ a redirect case?  Something appears fishy here.  (Whatever the resolution is for this test applies to the next test too.)

Reply:
Even if I mock `isRedirect` it still skips it? Maybe I'll remove the tests